docs/manpages/tracker-reset.1
.TH tracker-reset 1 "September 2014" GNU "User Commands"

.SH NAME
tracker-reset \- Reset the index and configuration

.SH SYNOPSIS
\fBtracker reset\fR [\-\-hard | \-\-soft] [\-\-config] [\-\-file \fIFILE\fR]

.SH DESCRIPTION
The reset command will change either your configuration or index
irreversibly and should be used with care. Other than tags, actual
data (e.g. files) should not be affected by this command.

The "index" is a link between your content (either locally or
remotely) and how it can be found quickly using a number of different
queries. Under the hood, this is done using a database.

Removing all data and starting again from the beginning with an empty
data set (which is a common use of this command) is done by using the
hard reset option. This behaves as if Tracker was just installed.

.SH OPTIONS
.TP
.B \-r, \-\-hard
This kills all processes in the same way that \fBtracker daemon
\-\-kill\fR does but it also removes all databases. Restarting
\fBtracker-store\fR re-creates the databases.
.TP
.B \-e, \-\-soft
A soft reset works exactly the same way that \fB\-\-hard\fR does, with
the exception that the backup and journal are not removed. These are
restored when \fBtracker-store\fR is restarted. This command is useful if
you have a corrupt database but want to reply the journal to restore
it to the last known good place.
.TP
.B \-c, \-\-config
Resets all configuration to its defaults.
.TP
.B \-f, \-\-file \fIFILE\fR
Resets all indexed information about \fIFILE\fR, works recursively for
directories. Nothing will be done if \fIFILE\fR is not currently indexed.
After deletion, a request to reindex this data will be immediately issued.

.SH SEE ALSO
.BR tracker-daemon (1).
.BR tracker (1).


NEWS
NEW in 1.8.0 - 2016-03-21
=========================

Changes since 1.7.5:
  * libtracker-miner: Adapt to libcue 2.0
  * translations: da

Overview of changes since 1.6.x:
  * Better support of the Sparql 1.1 syntax:
    - Support for many builtin functions:
      * String: CONCAT, CONTAINS, LCASE, UCASE, STRLEN, SUBSTR, STRSTARTS, STRENDS,
        STRBEFORE, STRAFTER, ENCODE_FOR_URI.
      * Numeric: ABS, ROUND, CEIL, FLOOR, RAND.
      * Date/Time: YEAR, MONTH, DAY, HOUR, MINUTES, SECONDS, NOW.
      * Checksum: MD5, SHA1, SHA256, SHA512.
  * Support for Sparql 1.1's DELETE {} INSERT {} WHERE {} syntax. This allows for
    performing complex updates in a single operation, without the shortcomings of
    the tracker specific INSERT OR REPLACE {} syntax.
  * Support for Sparql 1.1's BIND(). This allows mapping complex evaluations to a
    variable within the triple pattern, so usable within it.
  * The latter two applied allowed tracker-miner-fs to halve the time spent in
    handling move operations.
  * Many fixes to tracker-miner-fs and libtracker-miner, it is now a lot more robust
    esp. around cancellation situations (eg. deleting/unmounting the folder currently
    being indexed).
  * Many fixes to tracker-extract modules, many consistency fixes have been fixed where
    modules would produce invalid Sparql (eg. breaking cardinality limits of certain
    variables) if given files with unexpected metadata.
  * As a result of the last two points, 1.8.0 fixes most of the warnings/criticals that
    have been reported as "spamming journald".
  * Tracker now uses GTask, except for vala code.

NEW in 1.7.5 - 2016-03-14
=========================

  * Add back .trackerignore match to ignored-directories-with-content
  * tracker-extract: Fix gstreamer module cuesheet handling
  * libtracker-miner: Avoid querying file type in crawling queries
  * libtracker-data: Handle inserts where the subproperty cardinality is larger than the parents'
  * tracker-extract: Protect all single-valued properties in abiword extractor
  * tracker-extract: Protect all single-valued properties in EPUB extractor
  * tracker-extract: Protect all single-valued properties in ooxml extractor
  * tracker-extract: Protect all single-valued properties in oasis extractor
  * tracker-extract: Protect all single-valued properties in HTML extractor
  * tracker-extract: Check string length before parsing XMP in PDF extractor
  * tracker-extract: Add missing application/msword mimetype
  * libtracker-miner: Cut some slack on the reentry counter
  * libtracker-miner: Avoid changing order of elements in processing queue
  * libtracker-extract: Delete TrackerExtractClient
  * tracker-extract: Remove old dbus interface xml
  * tracker-extract: propagate urn to the TrackerExtractInfo
  * tracker-extract: Use safer method to insert tags in PDF module
  * tracker-extract: Use safer method to insert tags in GIF module
  * tracker-extract: Use safer method to insert tags in JPEG module
  * tracker-extract: Use safer method to insert tags in TIFF module
  * tracker-extract: Use safer method to insert tags in PNG module
  * libtracker-miner: Initialize all NodeData memory
  * libtracker-miner: Ensure the directory root is removed when its indexing root is

NEW in 1.7.4 - 2016-03-01
=========================

  * Update to FTS5.
  * libtracker-miner: Many fixes to TrackerFileNotifier cancellation
  * libtracker-direct: Handle cancellable argument in queries
  * libtracker-miner: Plug fd leak on TrackerCrawler cancellation
  * libtracker-extract: Fix year-only date extraction in gstreamer module
  * libtracker-extract: Use tracker-guarantee to ensure a title in playlists
  * tracker-miner-fs: Ignore git repositories. Modify the ignored-directories-with-content
      setting if you found this convenient.
  * tracker-miner-fs: Ignore #*# vim backups

Translations: oc

NEW in 1.7.3 - 2016-02-16
=========================

  * libtracker-miner: Many indexing fixes leading to stale elements in the database.
      If "UNIQUE constraint failed: nie:DataObject.nie:url" errors are seen in
      journald, running tracker-miner-fs once with the TRACKER_MINER_FORCE_CHECK_UPDATED
      envvar is recommended. you will need to terminate miners before that with
      tracker daemon -t
  * libtracker-miner: Do not insert partial/empty sparql on error
  * libtracker-miner: Pass a builder in UPDATE state to TrackerMinerFS::remove-file
  * libtracker-miner: Remove children recursively from queues on directory deleted
  * libtracker-miner: Fix generated Sparql query in sparql_contents_compose_query()
  * libtracker-miner: Fix some memory leaks of TrackerTask
  * libtracker-miner: Invalidate files iri recursively in case of file removal
  * libtracker-miner: Reset of reentry counter is not needed anymore
  * libtracker-fts: Fix invalid blob length calculation
  * libtracker-common: Use guint64 for free space calculations
  * libtracker-data, docs, libtracker-miner: Fix compile warnings
  * libtracker-data: misc code fixes
  * libtracker-data: Fix g_warning() missing argument
  * Update AppData to spec version 0.7+

Translations: lv

NEW in 1.7.2 - 2016-01-17
=========================

  * Many leak fixes.
  * libtracker-data: Reverted code to clean up stale Resources, can't just be done yet.
  * tracker tool: Removed tracker-compatibility CLI wrapper for older commands.
  * libtracker-common: Fix possible warnings on libicu unaccent code
  * ontology: Set domain index on nie:contentCreated for nmo:Message
  * libtracker-miner: Add ::remove-file signal vfunc
  * libtracker-common: Return total available space if running as admin.

Translations: lt

NEW in 1.7.1 - 2015-12-17
=========================

  * libtracker-data: Support for Sparql 1.1 BIND and
    DELETE {...} INSERT {...} WHERE {...} syntax.
  * libtracker-data: Support for Sparql 1.1 functions: NOW, RAND
  * libtracker-miner: Improved performance of move operations in TrackerMinerFS

src/tracker/tracker-reset.c
/*
 * Copyright (C) 2014, Lanedo <martyn@lanedo.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include "config.h"

#include <stdlib.h>
#include <stdio.h>

#include <glib.h>
#include <glib/gi18n.h>
#include <glib/gprintf.h>
#include <gio/gio.h>

#include <libtracker-common/tracker-common.h>
#include <libtracker-data/tracker-data.h>
#include <libtracker-control/tracker-control.h>

#include "tracker-reset.h"
#include "tracker-daemon.h"
#include "tracker-process.h"
#include "tracker-config.h"
#include "tracker-color.h"

static gboolean hard_reset;
static gboolean soft_reset;
static gboolean remove_config;
static gchar *filename = NULL;

#define RESET_OPTIONS_ENABLED() \
	(hard_reset || \
	 soft_reset || \
	 remove_config || \
	 filename)

static GOptionEntry entries[] = {
	{ "hard", 'r', 0, G_OPTION_ARG_NONE, &hard_reset,
	  N_("Kill all Tracker processes and remove all databases"),
	  NULL },
	{ "soft", 'e', 0, G_OPTION_ARG_NONE, &soft_reset,
	  N_("Same as --hard but the backup & journal are restored after restart"),
	  NULL },
	{ "config", 'c', 0, G_OPTION_ARG_NONE, &remove_config,
	  N_("Remove all configuration files so they are re-generated on next start"),
	  NULL },
	{ "file", 'f', 0, G_OPTION_ARG_FILENAME, &filename,
	  N_("Erase indexed information about a file, works recursively for directories"),
	  N_("FILE"),
	  NULL },
	{ NULL }
};

static void
log_handler (const gchar    *domain,
             GLogLevelFlags  log_level,
             const gchar    *message,
             gpointer        user_data)
{
	switch (log_level) {
	case G_LOG_LEVEL_WARNING:
	case G_LOG_LEVEL_CRITICAL:
	case G_LOG_LEVEL_ERROR:
	case G_LOG_FLAG_RECURSION:
	case G_LOG_FLAG_FATAL:
		g_fprintf (stderr, "%s\n", message);
		fflush (stderr);
		break;
	case G_LOG_LEVEL_MESSAGE:
	case G_LOG_LEVEL_INFO:
	case G_LOG_LEVEL_DEBUG:
	case G_LOG_LEVEL_MASK:
	default:
		g_fprintf (stdout, "%s\n", message);
		fflush (stdout);
		break;
	}
}

static void
delete_file (GFile    *file,
             gpointer  user_data)
{
	if (g_file_delete (file, NULL, NULL)) {
		gchar *path;

		path = g_file_get_path (file);
		g_print ("  %s\n", path);
		g_free (path);
	}
}

static void
directory_foreach (GFile    *file,
                   gchar    *suffix,
                   GFunc     func,
                   gpointer  user_data)
{
	GFileEnumerator *enumerator;
	GFileInfo *info;
	GFile *child;

	enumerator = g_file_enumerate_children (file, G_FILE_ATTRIBUTE_STANDARD_NAME,
	                                        G_FILE_QUERY_INFO_NONE, NULL, NULL);

	while ((info = g_file_enumerator_next_file (enumerator, NULL, NULL)) != NULL) {

		if (!suffix || g_str_has_suffix (g_file_info_get_name (info), suffix)) {
			child = g_file_enumerator_get_child (enumerator, info);
			(func) (child, user_data);
			g_object_unref (child);
		}

		g_object_unref (info);
	}

	g_object_unref (enumerator);
}

static int
delete_info_recursively (GFile *file)
{
	TrackerSparqlConnection *connection;
	TrackerMinerManager *miner_manager;
	TrackerSparqlCursor *cursor;
	gchar *query, *uri;
	GError *error = NULL;

	connection = tracker_sparql_connection_get (NULL, &error);

	if (error)
		goto error;

	uri = g_file_get_uri (file);

	/* First, query whether the item exists */
	query = g_strdup_printf ("SELECT ?u { ?u nie:url '%s' }", uri);
	cursor = tracker_sparql_connection_query (connection, query,
	                                          NULL, &error);

	/* If the item doesn't exist, bail out. */
	if (!cursor || !tracker_sparql_cursor_next (cursor, NULL, &error)) {
		g_clear_object (&cursor);

		if (error)
			goto error;

		return EXIT_SUCCESS;
	}

	g_object_unref (cursor);

	/* Now, delete the element recursively */
	g_print ("%s\n", _("Deleting…"));
	query = g_strdup_printf ("DELETE { "
	                         "  ?f a rdfs:Resource . "
	                         "  ?ie a rdfs:Resource "
	                         "} WHERE {"
	                         "  ?f nie:url ?url . "
	                         "  ?ie nie:isStoredAs ?f . "
	                         "  FILTER (?url = '%s' ||"
	                         "          STRSTARTS (?url, '%s/'))"
	                         "}", uri, uri);
	g_free (uri);

	tracker_sparql_connection_update (connection, query,
	                                  G_PRIORITY_DEFAULT, NULL, &error);
	g_free (query);

	if (error)
		goto error;

	g_object_unref (connection);

	g_print ("%s\n", _("The indexed data for this file has been deleted "
	                   "and will be reindexed again."));

	/* Request reindexing of this data, it was previously in the store. */
	miner_manager = tracker_miner_manager_new_full (FALSE, NULL);
	tracker_miner_manager_index_file (miner_manager, file, &error);
	g_object_unref (miner_manager);

	if (error)
		goto error;

	return EXIT_SUCCESS;

error:
	g_warning ("%s", error->message);
	g_error_free (error);
	return EXIT_FAILURE;
}

static gint
reset_run (void)
{
	GError *error = NULL;

	if (hard_reset && soft_reset) {
		/* TRANSLATORS: --hard and --soft are commandline arguments */
		g_printerr ("%s\n",
		            _("You can not use the --hard and --soft arguments together"));
		return EXIT_FAILURE;
	}

	if (hard_reset || soft_reset) {
		gchar response[100] = { 0 };

		g_print (CRIT_BEGIN "%s" CRIT_END "\n%s\n\n%s %s: ",
		         _("CAUTION: This process may irreversibly delete data."),
		         _("Although most content indexed by Tracker can "
		           "be safely reindexed, it can't be assured that "
		           "this is the case for all data. Be aware that "
		           "you may be incurring in a data loss situation, "
		           "proceed at your own risk."),
		         _("Are you sure you want to proceed?"),
		         /* TRANSLATORS: This is to be displayed on command line output */
		         _("[y|N]"));

		fgets (response, 100, stdin);
		response[strlen (response) - 1] = '\0';

		/* TRANSLATORS: this is our test for a [y|N] question in the command line.
		 * A partial or full match will be considered an affirmative answer,
		 * it is intentionally lowercase, so please keep it like this.
		 */
		if (!response[0] || !g_str_has_prefix (_("yes"), response)) {
			return EXIT_FAILURE;
		}
	}

	if (filename) {
		GFile *file;
		gint retval;

		file = g_file_new_for_commandline_arg (filename);
		retval = delete_info_recursively (file);
		g_object_unref (file);
		return retval;
	}

	/* KILL processes first... */
	if (hard_reset || soft_reset) {
		tracker_process_stop (TRACKER_PROCESS_TYPE_NONE, TRACKER_PROCESS_TYPE_ALL);
	}

	if (hard_reset || soft_reset) {
		guint log_handler_id;
#ifndef DISABLE_JOURNAL
		gchar *rotate_to;
		TrackerDBConfig *db_config;
		gsize chunk_size;
		gint chunk_size_mb;
#endif /* DISABLE_JOURNAL */

		/* Set log handler for library messages */
		log_handler_id = g_log_set_handler (NULL,
		                                    G_LOG_LEVEL_MASK | G_LOG_FLAG_FATAL,
		                                    log_handler,
		                                    NULL);

		g_log_set_default_handler (log_handler, NULL);

#ifndef DISABLE_JOURNAL
		db_config = tracker_db_config_new ();

		chunk_size_mb = tracker_db_config_get_journal_chunk_size (db_config);
		chunk_size = (gsize) ((gsize) chunk_size_mb * (gsize) 1024 * (gsize) 1024);
		rotate_to = tracker_db_config_get_journal_rotate_destination (db_config);

		/* This call is needed to set the journal's filename */
		tracker_db_journal_set_rotating ((chunk_size_mb != -1),
		                                 chunk_size, rotate_to);

		g_free (rotate_to);
		g_object_unref (db_config);

#endif /* DISABLE_JOURNAL */

		/* Clean up (select_cache_size and update_cache_size don't matter here) */
		if (!tracker_db_manager_init (TRACKER_DB_MANAGER_REMOVE_ALL,
		                              NULL,
		                              FALSE,
		                              FALSE,
		                              100,
		                              100,
		                              NULL,
		                              NULL,
		                              NULL,
		                              &error)) {

			g_message ("Error initializing database: %s", error->message);
			g_free (error);

			return EXIT_FAILURE;
		}
#ifndef DISABLE_JOURNAL
		tracker_db_journal_init (NULL, FALSE, NULL);
#endif /* DISABLE_JOURNAL */

		tracker_db_manager_remove_all (hard_reset);
		tracker_db_manager_shutdown ();
#ifndef DISABLE_JOURNAL
		tracker_db_journal_shutdown (NULL);
#endif /* DISABLE_JOURNAL */

		/* Unset log handler */
		g_log_remove_handler (NULL, log_handler_id);

		if (!remove_config) {
			return EXIT_SUCCESS;
		}
	}

	if (remove_config) {
		GFile *file;
		const gchar *home_conf_dir;
		gchar *path;
		GSList *all, *l;

		/* Check the default XDG_DATA_HOME location */
		home_conf_dir = g_getenv ("XDG_CONFIG_HOME");

		if (home_conf_dir && tracker_path_has_write_access_or_was_created (home_conf_dir)) {
			path = g_build_path (G_DIR_SEPARATOR_S, home_conf_dir, "tracker", NULL);
		} else {
			home_conf_dir = g_getenv ("HOME");

			if (!home_conf_dir || !tracker_path_has_write_access_or_was_created (home_conf_dir)) {
				home_conf_dir = g_get_home_dir ();
			}
			path = g_build_path (G_DIR_SEPARATOR_S, home_conf_dir, ".config", "tracker", NULL);
		}

		file = g_file_new_for_path (path);
		g_free (path);

		g_print ("%s\n", _("Removing configuration files…"));

		directory_foreach (file, ".cfg", (GFunc) delete_file, NULL);
		g_object_unref (file);

		g_print ("%s\n", _("Resetting existing configuration…"));

		all = tracker_gsettings_get_all (NULL);

		if (!all) {
			return EXIT_FAILURE;
		}

		for (l = all; l; l = l->next) {
			ComponentGSettings *c = l->data;
			gchar **keys, **p;

			if (!c) {
				continue;
			}

			g_print ("  %s\n", c->name);

			keys = g_settings_list_keys (c->settings);
			for (p = keys; p && *p; p++) {
				g_print ("    %s\n", *p);
				g_settings_reset (c->settings, *p);
			}

			if (keys) {
				g_strfreev (keys);
			}

			g_settings_apply (c->settings);
		}

		g_settings_sync ();

		tracker_gsettings_free (all);

		return EXIT_SUCCESS;
	}

	/* All known options have their own exit points */
	g_warn_if_reached ();

	return EXIT_FAILURE;
}

static int
reset_run_default (void)
{
	GOptionContext *context;
	gchar *help;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);
	help = g_option_context_get_help (context, FALSE, NULL);
	g_option_context_free (context);
	g_printerr ("%s\n", help);
	g_free (help);

	return EXIT_FAILURE;
}

static gboolean
reset_options_enabled (void)
{
	return RESET_OPTIONS_ENABLED ();
}

int
tracker_reset (int argc, const char **argv)
{
	GOptionContext *context;
	GError *error = NULL;

	context = g_option_context_new (NULL);
	g_option_context_add_main_entries (context, entries, NULL);

	argv[0] = "tracker reset";

	if (!g_option_context_parse (context, &argc, (char***) &argv, &error)) {
		g_printerr ("%s, %s\n", _("Unrecognized options"), error->message);
		g_error_free (error);
		g_option_context_free (context);
		return EXIT_FAILURE;
	}

	g_option_context_free (context);

	if (reset_options_enabled ()) {
		return reset_run ();
	}

	return reset_run_default ();
}
